### ISSUE TITLE
Add support for numpy.nonzero() to ndarray

### ISSUE THREAD

**Issue by ifeherva** [apache/incubator-mxnet#11424]
The [numpy.nonzero()](https://docs.scipy.org/doc/numpy-1.14.0/reference/generated/numpy.nonzero.html) operator is currently missing though it is very useful.

By providing a native implementation of nonzero in mxnet NDarray

- nd.where() can be extended to mimic np.where(condition)
- np.choice() can be implemented fully on gpu
- and many more

**Comment by frankfliu** [comment]
Hi @ifeherva thanks for submitting the issue, @sandeep-krishnamurthy requesting this be labeled under Feature request

**Comment by djl11** [comment]
You can achieve this affect by creating a CSRNDArray, as explained [here](https://github.com/apache/incubator-mxnet/blob/master/docs/tutorials/sparse/csr.md).

If you have a 2D array which contains zeros, you can convert this to a csr array like so:
`some_array = some_array.tostype('csr')`

you can then extract the indices of the non-zero entries like so:
`some_array.indices`

This is not to say that a non-zero function directly in mx.nd wouldn't be useful, but this is a good current solution for anybody who is searching.

**Comment by ifeherva** [comment]
Thank you @djl11 this has really helped me in implementing unique, where and do on. However, only in nd. Is there a way to do this in symbolic?

**Comment by eric-haibin-lin** [comment]
@ifeherva unfortunately the .data/.indices APIs are only available in the imperative mode. @junrushao1994 is looking into dynamic shape operators and might have some idea of what is coming

**Comment by ifeherva** [comment]
@eric-haibin-lin the .indices API is a blocking call right? I am getting very slow speeds and that would explain it.

**Comment by eric-haibin-lin** [comment]
That's right. It is blocking because the shape of the output cannot be inferred until the data is seen. You can see the dynamic shape support proposal in https://cwiki.apache.org/confluence/display/MXNET/Dynamic+shape I am not aware of the target date for the non-blocking implementation.

**Comment by JustinhoCHN** [comment]
Any update here? I want to use np.where, np.nonzero or np.unique in mxnet ndarray version.

**Comment by Kh4L** [comment]
In addition, this operator should return a NDArray of shape `(number_of_nonzeros, num_dims)`, directly representing the indices, rather than a tuple of indices like in NumPy.
